### ISSUE TITLE
Add support to define widget preview and see it in the IDE when developing

### ISSUE THREAD

**Issue by mboron** [flutter/flutter#115704]
## Use case

It is difficult to quickly find out what widget contains the dart file when opening it in the IDE. A developer has to read the code and try to guess what the widget looks like and whether this is the widget he searches for. Because of that, it is hard to get oriented in a project with a large code base or when joining a project created by someone else, or simply when you don't remember widget names by heart. One has to run the app and search for a particular widget e.g. based on strings visible in the app UI.
Also, a developer has to run the app when developing even a simple widget.

## Proposal

I would like to see a widget preview(s) when opening a dart file. Also, I would like to be able to define some example data for the widget preview, or even more preview options like font scale.
That would allow me to develop widgets without a need to run the app and orient quickly in the project.
 
In general, it would be useful to have the preview functionality similar to:
	• Android Compose annotations - @Preview/@PreviewParameter
		- https://developer.android.com/jetpack/compose/tooling#preview
	• or Swift UI PreviewProvider
 
Btw: it seems that there is starting one project that is similar to what I aim for -  https://pub.dev/packages/idea_widget_preview

**Comment by konkontos** [comment]
I strongly second this one. 

I'm coming to Flutter from a UIKit/SwiftUI background and I'm amazed that  there's no first-party tool to preview widgets.
This is a **must-have** feature.

**Comment by TacticalTwerking** [comment]
totally agree, will be wasting a lot of time in file seeking at the current stage

**Comment by SiXuManYan** [comment]
I am also troubled by this problem, I still can't quickly preview the layout file like in the android project, please officially launch the weight preview tool as soon as possible

**Comment by mboron** [comment]
I think the best way to understand the use case is, e.g., to download the Android compose-samples project ([LINK](https://github.com/android/compose-samples)) and open the `JetNews` sample project in Android Studio.

Below is an Android Studio screenshot from the `JetNews` project.
Let's say I'm interested in the `PostCardPopular` UI component. 
I don't have to run the app and find where it is displayed in the app to see how the component looks to find out whether this is the component I searched for.
I can open the file and directly display the preview on the right side.
You can do the same for Android projects with XML-based layouts or iOS projects with Swift UI.
It would be very helpful to have such previews available also for Flutter development.

<img width="2032" alt="Screenshot 2023-08-01 at 23 00 27" src="https://github.com/flutter/flutter/assets/7016173/6091c002-7fa8-48b2-a3fb-da45cf0337d2">

**Comment by Hixie** [comment]
That sounds like a yes? That seems like a reasonable use case.
It looks like https://pub.dev/packages/idea_widget_preview is indeed working on that use case, is it sufficient for your needs or are there use cases it doesn't handle?

It does seem like it would be particularly nice to be able to do this without having to define an explicit Preview widget and without having to be on an IDE (e.g. it'd be nice to have this inline in GitHub while looking at a codebase).

**Comment by mboron** [comment]
Yes, you described the use case correctly - the idea is to navigate a codebase and quickly see for each `Widget` subclass what a typical instance of that widget looks like.

Unfortunately, the `idea_widget_preview` plugin was recently deprecated and removed from the JetBrains marketplace.
It is mentioned here: [https://github.com/andrasferenczi/idea_widget_preview](https://github.com/andrasferenczi/idea_widget_preview)

I'm currently not aware of any other similar solution for Android Studio.

Such functionality is essential for developers and its lack negatively influences developer experience and productivity.

**Comment by Hixie** [comment]
FWIW the author of that package says [they'd be happy](https://github.com/andrasferenczi/idea_widget_preview/commit/c94d351a5e12e4860a7f852ba59738ec127c6353#commitcomment-124557714) for someone to take it over, it sounds like they just moved on to other things, it wasn't a technical problem with the plugin that caused it to be abandoned. I would definitely encourage people to take them up on that offer if this is something that interests you.

I've also approached our VS Code plugin team about this. It sounds like there's a difficulty in that environment in that we'd have to use Flutter for web since VS Code is basically a web page and embedding native windows into that environment would be non-trivial. That would introduce some unfortunate latency.

I wonder how we could get previews into something like GitHub's UI. That would be amazing.

**Comment by Hixie** [comment]
(Assigning to myself to follow up on the idea and maybe write a design doc, not to implement it.)

**Comment by adifyr** [comment]
This feature would be a huge upgrade in developer experience. I miss Android's XML UI editor from the old days. Flutter felt like a step backward in this particular direction.

**Comment by goderbauer** [comment]
@reidbaker suggests that something could be built in an IDE that displays widgets defined in a widget test as an MVP.

**Comment by goderbauer** [comment]
cc @anderdobo @jacob314 FYI

**Comment by Hixie** [comment]
avoiding having to explicitly write code for this (like with swiftui/jpc) would be a huge win.

**Comment by ClementCardonnel** [comment]
> avoiding having to explicitly write code for this (like with swiftui/jpc) would be a huge win.

As a SwiftUI developer myself, I don't mind writing this code as it allows me to write configurations so my previews can reflect real use cases, be put in ScrollViews, inside constrained layouts, etc.

Just my opinion. Anyway, it would be cool if Flutter supported this feature indeed.

**Comment by manuthebyte** [comment]
Any update on this? Would love a feature like that.

**Comment by anderdobo** [comment]
Nothing to report yet, but our team _is_ making progress on our investigations for this feature.

**Comment by adifyr** [comment]
The fact that we can't see previews for the individual widgets we create is the biggest productivity hindrance to this framework.

So, unless we run our app (with all the data and piping necessary), navigate to our route, we don't get to see what the widget looks like. What a pain.

**Comment by reidbaker** [comment]
@adifyr have you tried widget tests. The documentation is here https://docs.flutter.dev/cookbook/testing/widget/introduction and it is specifcally designed for letting you test widgets independent of running an app.

**Comment by adifyr** [comment]
@reidbaker appreciate the reply. But, I'd like to see my widgets, not test them. 😭🤌

**Comment by reidbaker** [comment]
Widget tests have the ability to compare to a golden screen shot so you can see them.

**Comment by bkonyi** [comment]
@reidbaker while that's true, there's quite a bit more latency between making a change and viewing the golden screenshot.  

@adifyr, as mentioned by @anderdobo, this is currently being investigated. While we don't have anything concrete to report yet, we're currently working on a high level design for this functionality and will hopefully have more to share in the coming months.

**Comment by burhanrashid52** [comment]
Here are my thoughts from flutter forum on why we need this

So I was working on SwiftUI last week and found the `#Preview` feature which is similar to the `@Preview` feature in Jetpack Compose. So I thought it would be great to have this feature in Flutter in-build DevTool as well. Although we can argue that hot-reload is fast enough to make this all work, but here are the reasons I would love to have them:

1. **Prototype a small widget in isolation**: This is especially helpful in big projects where you need to design a small widget that is deeply embedded within a user workflow. In order to test with hot-reload, you need to go through that workflow and test it. There are tool like Storybooks, but then I would have to create all my reusable theme and style components to be there and also keep them up to date.

2. **Live test with different inherited models**: SwiftUI provides a .environment modifier which takes the model we want to use in our SwiftUI View. A similar idea could be used in Flutter for an InheritedWidget or a third-party package that supports it.

3. **Interaction in small widgets:** The #Preview feature in SwiftUI is interactive. I can test small UI changes easily in the preview itself. Like show/hide and toggles.

I am aware that there are some [third-party packages](https://pub.dev/packages/device_preview) available for similar things, but they require some additional Dart code to set up inside the app.

**Comment by anderdobo** [comment]
@burhanrashid52 thank you for your insights on the use case. As @bkonyi mentioned, we're actively working on this so stay tuned!

**Comment by bkonyi** [comment]
Status update: this is actively being worked on.

**Comment by manuthebyte** [comment]
Is this not it?

https://docs.flutter.dev/tools/widget-previewer

didnt know it was already available 🚀

**Comment by kenzieschmoll** [comment]
This is complete CC @bkonyi
